src/framework-ondata.js

```diff
@@ -1,18 +1,294 @@
+    var currentState = stateFacade();
+    currentState.data = newData;
+    if ( newLm ) { 
+        currentState.lm = createLm();
+    }
+    stateFacade( currentState );
+}
+
+/**
+ * Clears the state data field by setting it to undefined.  
+ *
+ * @param stateFacade a getter/setterfunction for the current state
+ */
+function clearStateData( stateFacade ) { 
+    changeStateData( stateFacade ); 
+}
+
+/**
+ * Sends an error state on to any attached port on a down stream node if it's a new error.
+ * If there is no attached port, logs the error so it can be analyzed.
+ *
+ * @param port the error port, which may or may not be attached to something down stream
+ * @param lastLm last recorded state lm
+ * @param state an error or output state to be sent down stream or logged
+ */
+function handleError( port, lastLm, state ) { 
+
+    if ( (! handleOutput( port, lastLm, state )) && state.data )  { 
+        // State was not sent on to an attached port, and we do have error data
+        // so go ahead and log it for debugging/support use.
+        if ( state.data.stack ) { 
+            console.error( state.data.stack );
+        } else { 
+            console.error( state.data );
+        }
+    }
+}
+
+/**
+ * Returns true if the state is an initial vni state that has not got any
+ * values in it yet.
+ */
+function isInitState( state ) { 
+    return ( _.isUndefined( state.data ) && _.isUndefined( state.lm ) && _.isUndefined( state.error ) );
+}
+
+/**
+ * Checks if state is new or not.  If it is new, it sends it down stream to any attached 
+ * nodes; if there are no attached nodes, the state data will be logged.
+ *
+ * @param port an output port, which may or may not be attached to something down stream
+ * @param lastLm last recorded state lm
+ * @param state an error or output state to be sent down stream or logged
+ *
+ * @return true if the data was sent to an attached port on a down stream node;
+ *         false if the data was not sent on
+ */
+function handleOutput( port, lastLm, state ) {
+
+    // Do we have a new state? 
+    if ( lastLm !== state.lm ) { 
+
+        // Got any edges out of this port? 
+        if (port.listAttached().length) {
+            port.send( state );
+            port.disconnect();
+            return true;
+        }  
+    } 
+
+    return false;
+}
+
+/**
+ * Set the output state error flag to true if there is a current error, false if not.
+ * 
+ * @param vni 
+ * @param errorFlag (optional) if not undefined, the errorFlag value will be set
+ */
+function setOutputErrorFlag( vni, errorFlag ) {
+
+    var outputState  = vni.outputState();
+    if ( _.isUndefined( errorFlag ) ) { 
+        outputState.error = ! _.isUndefined( vni.errorState().data );
+    } else { 
+        outputState.error = errorFlag;
+    }
+
+    if ( outputState.error ) { 
+         // If we have an error, we should forward the output so downstream nodes know about it
+         outputState.lm = createLm();
```

Comment:
No need to change this now, but I've logged an issue to change this sometime in the future.  The downstream nodes should be notified, but the outputState.lm should not be changed, for the reason explained more fully here:
https://github.com/rdf-pipeline/noflo-rdf-pipeline/issues/33